### PR TITLE
[pcl] Update to 1.12.0 [rtabmap] Update to 0.20.13

### LINKED ISSUE #15130: [pcl] Visualization headers are not installed
When pcl libraries are installed, I am trying to compile some samples found in the tutorials, but I cannot compile because the pcl/visualization headers cannot be found.

They do not exist in vcpkg/installed/include directory. Maybe they should be installed and the required components compiled.

### LINKED ISSUE #17662: [pcl] Qhull is disabled in pcl_config.h, causes pcl::ConvexHull to not compile or (with workaround) to link
**Host Environment**
- OS: Windows 10 Pro 1909
- Compiler: v142 (Visual Studio 2019)

**To Reproduce**

1. `.\vcpkg install pcl:x64-windows`
2. Examine `.\vcpkg\installed\x64-windows\include\pcl\pcl_config.h` for `/* #undef HAVE_QHULL */`. Expected: `#define HAVE_QHULL`

**Additional context**
Using yesterday's Revision 95836f4883ffb481b7dd5f89569bc6ef55c31a7e of Vcpkg the problem occurs. Two weeks ago it worked fine with 807a7987612bbd702be9d9b8069124be62aff857. In the meantime issue https://github.com/microsoft/vcpkg/issues/13881, that addressed a related problem, was resolved and PR https://github.com/microsoft/vcpkg/pull/17437 was merged to master. Maybe a new bug was introduced?

`.\vcpkg\installed\x64-windows\bin\pcl_surface.dll` does not export any `pcl::ConvexHull` symbols or anything with `hull`.

Creating a test project with the environment set to the `.\vcpkg\installed\x64-windows\include` and `.\vcpkg\installed\x64-windows\lib` directories, linking all *.lib files in the latter.
Naively using pcl::ConvexHull gives compiler errors:

```
// ConvexHull Test.cpp

#include "pcl/point_types.h"
#include "pcl/point_cloud.h"
#include "pcl/surface/convex_hull.h"

int main() {
    pcl::PointCloud<pcl::PointXYZ> pc;
    pcl::ConvexHull<pcl::PointXYZ> hull;

    hull.reconstruct(pc);
}
```
Build log:
```
1>C:\ConvexHull Test\ConvexHull Test.cpp(12,10): error C2039: 'ConvexHull': is not a member of 'pcl'
1>C:\ConvexHull Test\ConvexHull Test.cpp(12,20): error C2065: 'ConvexHull': undeclared identifier
1>C:\ConvexHull Test\ConvexHull Test.cpp(12,26): error C2275: 'pcl::PointXYZ': illegal use of this type as an expression
1>C:\ConvexHull Test\ConvexHull Test.cpp(12,36): error C2065: 'hull': undeclared identifier
1>C:\ConvexHull Test\ConvexHull Test.cpp(14,5): error C2065: 'hull': undeclared identifier
```

**Workaround** (kinda)
Manually defining `HAVE_QHULL` before including pcl_config.h (happens in any PCL header), compilation succeeds, but linking fails:
```
// ConvexHull Test.cpp

#define HAVE_QHULL

#include "pcl/point_types.h"
#include "pcl/point_cloud.h"
#include "pcl/surface/convex_hull.h"

int main() {
    pcl::PointCloud<pcl::PointXYZ> pc;
    pcl::ConvexHull<pcl::PointXYZ> hull;

    hull.reconstruct(pc);
}
```
Build log:
```
1>ConvexHull Test.obj : error LNK2001: unresolved external symbol "protected: virtual void __cdecl pcl::ConvexHull<struct pcl::PointXYZ>::performReconstruction(struct pcl::PolygonMesh &)" (?performReconstruction@?$ConvexHull@UPointXYZ@pcl@@@pcl@@MEAAXAEAUPolygonMesh@2@@Z)
1>ConvexHull Test.obj : error LNK2001: unresolved external symbol "protected: virtual void __cdecl pcl::ConvexHull<struct pcl::PointXYZ>::performReconstruction(class std::vector<struct pcl::Vertices,class std::allocator<struct pcl::Vertices> > &)" (?performReconstruction@?$ConvexHull@UPointXYZ@pcl@@@pcl@@MEAAXAEAV?$vector@UVertices@pcl@@V?$allocator@UVertices@pcl@@@std@@@std@@@Z)
1>ConvexHull Test.obj : error LNK2001: unresolved external symbol "public: void __cdecl pcl::ConvexHull<struct pcl::PointXYZ>::reconstruct(class pcl::PointCloud<struct pcl::PointXYZ> &)" (?reconstruct@?$ConvexHull@UPointXYZ@pcl@@@pcl@@QEAAXAEAV?$PointCloud@UPointXYZ@pcl@@@2@@Z)
1>C:\ConvexHull Test\x64\Release\ConvexHull Test.exe : fatal error LNK1120: 3 unresolved externals
```

### LINKED ISSUE #19016: [pcl] update to <1.12.0>
Library name: PCL ( Point Cloud Library ）

New version number: 1.12.0

Other information that may be useful (release notes, etc...)

### PR DESCRIPTION
**Pull request by raahilsha-z**
**Describe the pull request**

- #### What does your PR fix?  
Updates PCL to the latest 1.12.0

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
All currently supported triplets should still be supported, no changes needed to CI baseline

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
Yes updated

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

File: ports/pcl/portfile.cmake [hunk 1]
```diff
@@ -1,19 +1,18 @@
 vcpkg_from_github(
     OUT_SOURCE_PATH SOURCE_PATH
     REPO PointCloudLibrary/pcl
-    REF d98313133b014553ab1b1b5b112f9aade837d55c # pcl-1.11.1
-    SHA512 4d60f34d4fbf0a4b4caf9cc4391af471ebd260b3bbac106d45e5ff38448894ea4dc82d5320c2e395c537a4414eb13c6a6a0eb6c13e4e1cc1d831d4bf24067966
+    REF 8196d4d2534e8d5ffe62d4a4ea54d488618a661a # pcl-1.12.0-99
```
Comment (Neumann-A) [inline ports/pcl/portfile.cmake:4]: use f36a69a5e89953708990c4669317f989d532cf08 instead which is 1.12.0 ? .99 is inofficial post release. 
https://github.com/PointCloudLibrary/pcl/releases/tag/pcl-1.12.0 was release 5h ago from point of writing.

Comment (raahilsha-z) [reply ports/pcl/portfile.cmake:4]: Sorry about that! I'll push those changes ASAP

I also noticed that rtabmap seems to require pcl[vtk] to be built, and the CI is only building pcl[core]. Do I fix this by adding the dependency

{"name": "pcl", "features": ["vtk"]}

to rtabmap?

Comment (Neumann-A) [reply ports/pcl/portfile.cmake]: the latter if the other features are not required.

Comment (raahilsha-z) [reply ports/pcl/portfile.cmake]: Thank you! Gave that a shot and running `vcpkg install rtabmap:x64-windows` locally on my machine seems to automatically build `pcl[core,vtk]:x64-windows`. Just pushed it to the pull request

**Comment by raahilsha-z** [comment]
I'm unsure why the Linux and OSX checks are failing here, when they worked a couple commits ago: https://github.com/microsoft/vcpkg/pull/18855/checks?check_run_id=3012856022

The OSX one is failing to find a library called "m", and the Linux one is failing to find "GL"

**Comment by JonLiu1993** [comment]
@raahilsha-z ,Could you please take a look:
[config-x64-linux-dbg-err.log](https://github.com/microsoft/vcpkg/files/6782023/config-x64-linux-dbg-err.log)
[config-x64-linux-dbg-out.log](https://github.com/microsoft/vcpkg/files/6782024/config-x64-linux-dbg-out.log)

**Comment by raahilsha-z** [comment]
I just tried `vcpkg install rtabmap:x64-linux` on a fresh WSL2 Ubuntu 20.04 environment, and had no issues building PCL. I did have to run `apt-get install libxmu-dev libxi-dev libgl-dev` in order to get glew to compile, as per glew's portfile.cmake, and I see that glew and vtk are both succeeding inside of the CI environment, so I'm unsure why the `Library not found: GL` error is happening.

**Comment by JackBoosY** [comment]
> I've also noticed that rtabmap's manifest has "supports": "windows & !static" in it - does that imply that it shouldn't be built on Linux?

Yes, rtabmap only support Windows dynamic in vcpkg currently.

**Comment by JackBoosY** [comment]
```
CMake Error at /mnt/vcpkg-ci/installed/x64-linux/share/cmake/Qt5Gui/Qt5GuiConfig.cmake:95 (message):
  Library not found: GL
Call Stack (most recent call first):
  /mnt/vcpkg-ci/installed/x64-linux/share/cmake/Qt5Gui/Qt5GuiConfig.cmake:277 (_qt5_Gui_process_prl_file)
  /agent/_work/1/s/scripts/buildsystems/vcpkg.cmake:792 (_find_package)
  /mnt/vcpkg-ci/installed/x64-linux/share/cmake/Qt5Widgets/Qt5WidgetsConfig.cmake:231 (find_package)
  /agent/_work/1/s/scripts/buildsystems/vcpkg.cmake:792 (_find_package)
  /mnt/vcpkg-ci/installed/x64-linux/share/cmake/Qt5/Qt5Config.cmake:28 (find_package)
  /agent/_work/1/s/scripts/buildsystems/vcpkg.cmake:792 (_find_package)
  /mnt/vcpkg-ci/installed/x64-linux/share/vtk/VTK-vtk-module-find-packages.cmake:115 (find_package)
  /mnt/vcpkg-ci/installed/x64-linux/share/vtk/vtk-config.cmake:129 (include)
  /agent/_work/1/s/scripts/buildsystems/vcpkg.cmake:792 (_find_package)
  cmake/pcl_find_vtk.cmake:25 (find_package)
  CMakeLists.txt:380 (include)
```
Related to qt5-gui.
cc @Neumann-A

**Comment by larshg** [comment]
I don't think the PNG patch is required anymore, with the use of VTK 9 - the PNG are properly prefixed with optimized / debug and get linked accordingly.

**Comment by raahilsha-z** [comment]
Just tested removing the libpng patch and it seems to build fine on Windows 10 and Ubuntu 20.04

**Comment by Neumann-A** [comment]
> Related to qt5-gui. -> Error GL

requires `libglvnd` from the system? (Vendor neutral OpenGL)

**Comment by larshg** [comment]
The linemod targets compiles as well, so the remove-broken-targets.patch can be removed.

**Comment by larshg** [comment]
I just tried this PR on a Ubuntu 20.04.2 LTS.
When I let VCPKG do the cmake configuration using its cmake 3.20.2 it fails to find OpenGL (here cache is reloaded in 3.21.1):
image

However if I run the cmake manually with cmake 3.21.1 it does find it:
image

So can it be a Cmake version error, or is it vcpkg that somehow blocks cmake to properly find OpenGL?

**Comment by JonLiu1993** [comment]
Pinging @raahilsha-z  for response. Is work still being done for this PR?

**Comment by jasjuang** [comment]
Looking forward to seeing this getting completed. The current PCL 1.11 pretty much doesn't work with VTK 9. Really need PCL 1.12 for VTK 9 support.

**Comment by YhQIAO** [comment]
So i need to still waiting for the pcl 1.12 on vcpkg?

**Comment by raahilsha-z** [comment]
vcpkg currently uses cmake 3.21.1 now, and I've updated this fork to the current vcpkg head, but still runs into the same issues as when it used 3.20.2.

`CMake Error at /home/raahilsha/dev/vcpkg/downloads/tools/cmake-3.21.1-linux/cmake-3.21.1-linux-x86_64/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find OpenGL (missing: OPENGL_opengl_LIBRARY OPENGL_glx_LIBRARY)
Call Stack (most recent call first):
  /home/raahilsha/dev/vcpkg/downloads/tools/cmake-3.21.1-linux/cmake-3.21.1-linux-x86_64/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /home/raahilsha/dev/vcpkg/downloads/tools/cmake-3.21.1-linux/cmake-3.21.1-linux-x86_64/share/cmake-3.21/Modules/FindOpenGL.cmake:443 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  /home/raahilsha/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:786 (_find_package)
  cmake/pcl_find_gl.cmake:10 (find_package)
  CMakeLists.txt:418 (include)`

However, the library files exist exactly as they do in @larshg 's post above. I've tried searching a bit as to why vcpkg's cmake doesn't work while a separately downloaded one does, but so far it's been to no avail. I've also tried installing various other packages and using both a physical Ubuntu machine as well as a WSL2 one, but haven't found any difference between them.

Should the portfile be changed or a patch added somehow to find OpenGL in a different way? I'm not that familiar with CMake-based development on Linux. However, I'm able to install `qt5[3d,multimedia]` as well as `vtk` fine, which seem to also use OpenGL.

**Comment by jasjuang** [comment]
Updating PCL to 1.12 will resolve issues like https://github.com/PointCloudLibrary/pcl/issues/4946

I was able to build PCL 1.12 from source without errors by doing the below:

```
git clone https://github.com/PointCloudLibrary/pcl
git checkout pcl-1.12.0
cd pcl && mkdir build && cd build
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=/home/jasjuang/vcpkg/scripts/buildsystems/vcpkg.cmake ..
make -j
```

Looking forward to seeing PCL 1.12 in vcpkg.

**Comment by JackBoosY** [comment]
Ping @raahilsha-z  for response.

**Comment by raahilsha-z** [comment]
I think that jasjuang is seeing what larshg and I are seeing as well - calling CMake on your own on Ubuntu will let us build, but use of vcpkg to build PCL is leading to the GL libraries not being found. I'm not really sure why that's the case, especially considering that other vcpkg libraries that rely on OpenGL seem to build fine.

**Comment by JackBoosY** [comment]
@Neumann-A 
I found some thing interest:
```
qt5-base:/share/cmake/Qt5/Qt5Config.cmake
qt5-base:/share/cmake/Qt5/Qt5ConfigVersion.cmake
qt5-base:/share/cmake/Qt5/Qt5ModuleLocation.cmake
...
```
```
vcpkg@vcpkg % ls installed/x64-osx/share/cmake
Qt5                             Qt5Designer                     Qt5GraphicsSupport              Qt5PacketProtocol               Qt5QmlModels                    Qt5QuickWidgets                 Qt5UiTools
Qt5AccessibilitySupport         Qt5DesignerComponents           Qt5Gui                          Qt5PlatformCompositorSupport    Qt5QmlWorkerScript              Qt5ServiceSupport               Qt5Widgets
Qt5AttributionsScannerTools     Qt5DeviceDiscoverySupport       Qt5Help                         Qt5PrintSupport                 Qt5Quick                        Qt5Sql                          Qt5Xml
Qt5ClipboardSupport             Qt5EdidSupport                  Qt5LinguistTools                Qt5Qml                          Qt5QuickCompiler                Qt5Svg                          Qt5XmlPatterns
Qt5Concurrent                   Qt5EventDispatcherSupport       Qt5Network                      Qt5QmlDebug                     Qt5QuickParticles               Qt5Test                         ignition-cmake2
Qt5Core                         Qt5FbSupport                    Qt5OpenGL                       Qt5QmlDevTools                  Qt5QuickShapes                  Qt5ThemeSupport                 tl-expected
Qt5DBus                         Qt5FontDatabaseSupport          Qt5OpenGLExtensions             Qt5QmlImportScanner             Qt5QuickTest                    Qt5UiPlugin
```
Is this path incorrect?

**Comment by JackBoosY** [comment]
Since the regressions are related to qt5-base, we need another PR to fix that first.

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
Add `DISABLE_PARALLEL_CONFIGURE` to rtabmap to avoid configure error:
```
CMake Error at CMakeLists.txt:960 (CONFIGURE_FILE):
  CONFIGURE_FILE attempted to configure a file:
  D:/buildtrees/rtabmap/src/fde13e9126-d5e18ef027.clean/corelib/include/rtabmap/core/Version.h
  into a source directory.
```

**Comment by WangZP10969** [comment]
First of all, thank you very much for your efforts. I tried this version and it compiles normally under vs2019 (vs2022pre will have errors during pcl compilation, But I think this should be caused by 2022pre)

I encountered a problem in  #17662. I am not sure if there is a problem with my configuration method, but it seems that the performance of PCL1.12 and Qhull is similar to 1.11. HAVE_QHULL is not defined and part of the lib configuration is missing after manual definition.

I'm not sure if this is the cause of upstream or other problems, but just as a small hint.

**Comment by JJJohan** [comment]
> First of all, thank you very much for your efforts. I tried this version and it compiles normally under vs2019 (vs2022pre will have errors during pcl compilation [install-x64-windows-rel-out.log](https://github.com/microsoft/vcpkg/files/7461501/install-x64-windows-rel-out.log) , But I think this should be caused by 2022pre)
> 
> I encountered a problem in #17662. I am not sure if there is a problem with my configuration method, but it seems that the performance of PCL1.12 and Qhull is similar to 1.11. HAVE_QHULL is not defined and part of the lib configuration is missing after manual definition.
> 
> I'm not sure if this is the cause of upstream or other problems, but just as a small hint.

Hi WangZP, I ran into the same issue, as well as another one with cuda. I don't know what the proper fixes for these are, but I can provide some workarounds to get these building correctly (and linking as well 🙂)

These should work for both master as well as the raahilsha-z's branch.

For qlibhull:
In the vcpkg directory, open up the ports/pcl/portfile.cmake file and remove these two lines:
1. fix-find-qhull.patch
2. file(REMOVE ${SOURCE_PATH}/cmake/Modules/FindQhull.cmake)

That effectively reverts the PR changes mentioned in the #17662 issue and resolves the issue for me.

For CUDA:
This one's a bit of a dodgy workaround (yay!) as it involves editing the files after a build has been initiated (it worked for me, there's no doubt a more elegant way to do it. i.e. modifying the CMakeLists.txt files themselves and removing the cuda_common entry from the `SUBSYS_DEPS` variable)

1. Start a build (i.e. vcpkg install pcl[cuda]:x64-windows)
2. In the vcpkg directory, navigate to buildtrees/pcl/x64-windows-rel/cuda (as well as the x64-windows-dbg folder if required) and edit the .pc file found in the `features`, `sample_consensus` and `segmentation` folders, removing the mention of pcl_cuda_common-1.12.

It's a header-only module and doesn't produce the pkg-config files that it looks for after the build for patching / post-build related tasks which causes the whole thing to fail. It still seems to work fine afterwards for me.

**Comment by JackBoosY** [comment]
Ping @dan-shaw for review and merge this.

**Comment by JackBoosY** [comment]
> First of all, thank you very much for your efforts. I tried this version and it compiles normally under vs2019 (vs2022pre will have errors during pcl compilation [install-x64-windows-rel-out.log](https://github.com/microsoft/vcpkg/files/7461501/install-x64-windows-rel-out.log) , But I think this should be caused by 2022pre)
> 
> I encountered a problem in #17662. I am not sure if there is a problem with my configuration method, but it seems that the performance of PCL1.12 and Qhull is similar to 1.11. HAVE_QHULL is not defined and part of the lib configuration is missing after manual definition.
> 
> I'm not sure if this is the cause of upstream or other problems, but just as a small hint.

We already known this issue and reported it to the Visual Studio team.

**Comment by BillyONeal** [comment]
Thanks for the updates :)

**Comment by JackBoosY** [comment]
Okay, the pcl virtualization feature will not be blocked.

**Comment by AlexLeungZ** [comment]
I ran into another issue when installing pcl with features vtk

vcpkg.json
```
{
    "name": "robocon-lidar-pcl",
    "version": "1.2",
    "description": "LiDar dev with pcl for robocon.",
    "dependencies": [
        {
            "name": "pcl",
            "features": [ "vtk" ],
            "platform": "(linux & x64)"
        },
        {
            "name": "cuda",
            "features": [ ],
            "platform": "(linux & x64)"
        }
    ]
}
```

The issue is as same as this one: #21195 
The host environment, Steps to reproduce the behavior are as same as that one except for the vcpkg.json

Is there any way to use the pcl virtualization feature?

**Comment by AlexLeungZ** [comment]
> For CUDA: This one's a bit of a dodgy workaround (yay!) as it involves editing the files after a build has been initiated (it worked for me, there's no doubt a more elegant way to do it. i.e. modifying the CMakeLists.txt files themselves and removing the cuda_common entry from the `SUBSYS_DEPS` variable)
> 
> 1. Start a build (i.e. vcpkg install pcl[cuda]:x64-windows)
> 2. In the vcpkg directory, navigate to buildtrees/pcl/x64-windows-rel/cuda (as well as the x64-windows-dbg folder if required) and edit the .pc file found in the `features`, `sample_consensus` and `segmentation` folders, removing the mention of pcl_cuda_common-1.12.
> 
> It's a header-only module and doesn't produce the pkg-config files that it looks for after the build for patching / post-build related tasks which causes the whole thing to fail. It still seems to work fine afterwards for me.

Thanks for the idea of fixing pcl-cuda 🙂
I got a similar approach from yours which also successfully built for x64-linux

Here are the steps

1. create a new directory in your project directory (or wherever it is) with name `pkgconfig`
2. inside that directory, create a file called `pcl_cuda_common-1.12.pc`
3. `pcl_cuda_common-1.12.pc`
```
prefix=${pcfiledir}/../..

exec_prefix=${prefix}
libdir=${prefix}/lib
#includedir=${prefix}/include/pcl
includedir=${prefix}/include/pcl-1.12
Name: pcl_cuda_common-1.12
Description: 
Version: 1.12.0

Libs:  
Requires: 
Cflags: -I"${includedir}" 
```

4. goto `.../vcpkg/ports/pcl/portfile.cmake`
5. add this line `vcpkg_host_path_list(PREPEND ENV{PKG_CONFIG_PATH} "your/path/to/that/pkgconfig")` before `vcpkg_fixup_pkgconfig()`

I don't know what will be the consequence by doing that, but at least now the code will build.

**Comment by JackBoosY** [comment]
All the failures will be fixed in https://github.com/microsoft/vcpkg/pull/21276.

**Pull request merged**